src/coord/src/catalog/storage.rs

```diff
@@ -961,6 +1055,24 @@ struct DatabaseKey {
 }
 impl_codec!(DatabaseKey);
 
+#[derive(Clone, Message, PartialOrd, PartialEq, Eq, Ord, Hash)]
+struct ComputeInstanceReplicaKey {
+    #[prost(int64)]
+    compute_instance_id: i64,
+    #[prost(string)]
+    name: String,
+}
+impl_codec!(ComputeInstanceReplicaKey);
+
+#[derive(Clone, Message, PartialOrd, PartialEq, Eq, Ord)]
+struct ComputeInstanceReplicaValue {
+    #[prost(string)]
+    config: String,
+    #[prost(string)]
```

Comment:
What I was trying to say is that for `compute_instance_replicas`, I don't think we need a a unique value check. Stash already has a unique key constraint, which in this case will check that same thing as the unique value check. Which also means that we probably don't need to persist `service_name` at all.

Reply:
Ah, this code is actually meaningfully exercised:

1. [`sequence_create_compute_instance_replica`](https://github.com/MaterializeInc/materialize/pull/12132/files#diff-9fc9aab48f6290b549b1d95ea71e6aa4225ac9fa870a8a0b8aba29108694291aR1970) generates the replica service name out of bounds, using the same function that determines the service name in the lower layers. This must be done here pre-catalog transaction because we cannot permit creating two replicas with the same service name.
- [`DataflowTypes::Client::Controller::add_replica_to_instance`](https://github.com/MaterializeInc/materialize/pull/12132/files#diff-9d19deda2eb11325e9505642a623ba52f9084383939ddef679283425a4fc8e57R189) uses this same function to generate the service name, which we've guaranteed will be unique during sequencing.

> However since service_name is made up of the compute name and the replica name, there's a 1-to-1 mapping of compute name to compute instance id, the ComputeInstanceReplicaKey is made up of compute instance id and replica name

Ah, I think you've misunderstood the code in `generate_replica_service_name`. For instance `"foo_replica"` and `"foo-replica"` are both valid replica identifiers but generate the same service name (unless we wanted to add a nonce to the service names names, which is legit, but not what the code does) because e.g. underscores are not valid in hostnames, so we just blithely replace them with dashes. We then have to ensure that we haven't generated multiple replicas of the same name, which is what all of the song and dance with the catalog checks does.

Reply:
I can add some comments to the code to express this relationship more clearly to future readers--sorry that this wasn't more obvious.

Reply:
Ahh, ok that makes sense. Thanks for the explanation. In terms of the nonce I don't have a strong opinion, but I think it's probably not necessary for this PR.

Reply:
@sploiselle Just had a random thought this morning. I don't think it's necessary for this PR, just more food for thought. If we added disallowed characters to replica names s.t. it matched all the changes done by `sanitize_str_for_service_hostname`, then I think we could probably remove `service_name` from here and remove the `sanitize_str_for_service_hostname`.